---
title: "rarefaction"
output: html_document
date: "2025-05-23"
---

# Packages

```{r}
library(tidyverse)
library(vegan)
library(ape)
```


```{r}
data_18S <- "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2"
amf_taxonomic_levels_eucaryome <- c("kingdom", "phylum", "class", "order", "family",
                          "genus", "species")

amf_OTUs <- data_18S %>%
  #{{import data}}
  read_tsv() %>%
  #{{format column names}}
  rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~sub("_AMF_", "", .)) %>%
  rename_with(~sub("\\_AMF.*", "", .)) %>%
  rename_with(~sub("ROB_T_", "CONT_ROB_", .)) %>%
  rename_with(~sub("ROB_Tex", "CONT_ROB_T_ex", .)) %>%
  mutate(taxonomy = str_replace_all(taxonomy, ":","_"),
       taxonomy = str_replace_all(taxonomy, "[|]","; ")) %>%
  #{{select samples from ROBUST project}}
  dplyr::select(OTU, amplicon, taxonomy, starts_with("CONT", ignore.case = FALSE),
         starts_with("ROB", ignore.case = FALSE)) %>%
    #{{"Arbuscular mycorrhizal fungi & Mucoromycotina fine root endophytes" are selected}}
  filter(str_detect(taxonomy,"p__Glomeromycota")) %>%
  #filter(str_detect(taxonomy,"p__Glomeromycota")) %>%
  mutate(taxonomy = str_replace_all(taxonomy, "(p__|c__|o__|f__|g__|s__)unclassified", "*"))%>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F) %>%
  #{{select only OTUs and controls}}
  droplevels()

taxonomy <- amf_OTUs %>%
  dplyr::select(amplicon, OTU, taxonomy) %>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F)
 


d <- amf_OTUs %>% 
  replace(amf_OTUs == 0, NA) %>%
  dplyr::select(amplicon, starts_with("CONT")) %>%
  pivot_longer(-amplicon, names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads))

amf_OTUs_decont <- amf_OTUs %>%
  replace(amf_OTUs == 0, NA) %>%
  pivot_longer(starts_with("ROB"), names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads)) %>%
  #{{merge with control samples}}
  left_join(count(d, amplicon, wt = reads), by = "amplicon") %>%
  #{{substract abundance of control samples}}
  mutate(reads = case_when(
    is.na(n)  ~ reads,
    n > reads ~ 0,
    TRUE      ~ reads - n)) %>%
  dplyr::select(-n,-starts_with("CONT")) %>%
  pivot_wider(names_from = samples, values_from =  reads, values_fill = 0) %>%
  as_tibble() %>%
  column_to_rownames(var = "amplicon") 
```


# Rarefaction

## Rarefaction threshold

```{r}
plot.OTUcumSums <-function(x){plot(1:length(colSums(x)), log10(sort(colSums(x))+1))}
amf_OTUs_decont %>%
  dplyr::select(starts_with("ROB")) %>%
  plot.OTUcumSums
10^3.1
log10(10)
amf_OTUs_decont %>%
  dplyr::select(starts_with("ROB"))%>%
  colSums()%>%
  sort()


amf_OTUs_decont2 <- amf_OTUs_decont %>%
  #{{sample to exclude}}
  dplyr::select(starts_with("ROB"))%>%
  dplyr::select(where(~ sum(.x) >= 10^3.1))
smallest_amf <- min(colSums(amf_OTUs_decont2))
```

## rarefaction function

```{r}
rarefaction.func <- function(df,rar_sample, n_rare = 100, rarcur = T){
  # Function that rarefy "n_rare" times and average abundance across all rarefaction
  
  # If rarefaction curve needed
  if(rarcur){
    df %>%
      dplyr::select(starts_with("ROB")) %>%
      t() %>%
      vegan::rarecurve(step = 20, sample = rar_sample, col = "blue", cex = 0.6)
  }
  df %>%
    dplyr::select(starts_with("ROB")) %>%
    dim -> D
  
  # Compute rarefactions
  df_rarefy = matrix(0, nrow =  D[1], ncol = D[2])
  for(n in 1:n_rare){
    df %>%
      dplyr::select(starts_with("ROB")) %>%
      t() %>%
      vegan::rrarefy(rar_sample) %>%
      t() + df_rarefy -> df_rarefy
    
  }
  
  # Average and format output dataframe
  round(df_rarefy/n_rare) -> df_rarefy
  df_rarefy %>%
    bind_cols(
      df %>%
        dplyr::select(!starts_with("ROB"))
    ) -> df_rarefy
  rownames(df_rarefy) = rownames(df)
  return(df_rarefy)
}
```


```{r}
amf_OTUs_decont2 %>% 
  mutate(across(starts_with("ROB"), ~as.integer(.x))) -> amf_OTUs_decont2

set.seed(2121349)

rarefaction.func(df = amf_OTUs_decont2, rar_sample = smallest_amf, rarcur = T) -> amf_OTUs_rare

amf_OTUs_rare %>%
  as.data.frame()%>%
  rownames_to_column("amplicon") %>%
  left_join(amf_OTUs_decont %>%
              rownames_to_column("amplicon") %>%
              dplyr::select(-starts_with("ROB")), by ="amplicon") -> amf_OTUs_rare

write.table(amf_OTUs_rare, "outputs/abundance_tables/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.rarefy.txt")
```

# Abundance at genus level

```{r}
amf_OTUs_decont %>%
  #{{sample to exclude}}
  dplyr::select(where(~ ifelse(is.numeric(.x), sum(.x) >= 10^3.1, is.character(.x))))  %>%
  unite(taxonomy, kingdom, phylum, class, order, family, genus, sep = "; ") %>%
  dplyr::select(taxonomy, starts_with("ROB")) %>%
  group_by(taxonomy)%>%
  mutate(
    genus = paste(str_extract(
      taxonomy,"(g_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"), # If family isn't known agglomerate a the lower taxonomic level known
      n(),sep="_")
  ) %>%
  ungroup() %>%
  dplyr::select(-taxonomy) %>%
  group_by(genus) %>%
  summarise_all(sum) -> amf_genus_decont2


amf_genus_decont2 %>%
  column_to_rownames("genus") %>% colSums()

write.table(amf_genus_decont2, "outputs/abundance_tables/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.amf_genus_decont2.txt")
```

```{r}
amf_OTUs_rare %>% 
  unite(taxonomy, kingdom, phylum, class, order, family, genus, sep = "; ") %>%
  dplyr::select(taxonomy, starts_with("ROB")) %>%
  group_by(taxonomy)%>%
  mutate(
    genus = paste(str_extract(
      taxonomy,"(g_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"), # If family isn't known agglomerate a the lower taxonomic level known
      n(),sep="_")
  ) %>%
  ungroup() %>%
  dplyr::select(-taxonomy) %>%
  group_by(genus) %>%
  summarise_all(sum) -> amf_genus_rare


write.table(amf_genus_rare, "outputs/abundance_tables/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.afm_rarefy_allgenus.txt")
```


# Algomerate phylogenetic tree at the Genus level

```{r}
cluster_genus <- amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, genus, sep = "; ",remove = F) %>%
  select(taxonomy, amplicon, OTU)
```

```{r}
tree_amf <- read.tree(file = "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples_Eukaryome_Glomeromycota.tree")
plot(tree_amf, show.tips = F)
tree_amf
```

```{r}
# 1. Compute full pairwise patristic distance matrix between OTUs
dist_mat <- cophenetic.phylo(tree_amf)  # matrix of tip-to-tip distances

# 2. Prepare cluster mapping
otu_genus <- cluster_genus %>%
  # if species isn't known agglomerate a the lower taxonomic level known
  group_by(taxonomy) %>%
  mutate(
    genus = paste(str_extract(
      taxonomy,"(s_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  select(-taxonomy)

# 3. For each pair of species, compute median inter-cluster distance
genus_list <- unique(otu_genus$genus)

# Create a matrix to store median distances
genus_dist <- matrix(NA, nrow = length(genus_list), ncol = length(genus_list),
                       dimnames = list(genus_list, genus_list))

for (i in seq_along(genus_list)) {
  for (j in seq_along(genus_list)) {
    if (i <= j) {
      otus1 <- otu_genus$amplicon[otu_genus$genus == genus_list[i]]
      otus2 <- otu_genus$amplicon[otu_genus$genus == genus_list[j]]

      dists <- dist_mat[otus1, otus2, drop = FALSE]
      genus_dist[i, j] <- genus_dist[j, i] <- median(dists, na.rm = TRUE)
    }
  }
}

# 4. Create a new tree using median distances â€” convert matrix to phylo
genus_tree <- nj(as.dist(genus_dist))

# 5. Plot the collapsed tree
plot(genus_tree, cex = 0.7, main = "family-level Tree (collapsed from OTUs)")


ape::write.tree(phy = genus_tree, file = "outputs/phylo_tree/Robust_MiSeq_18S_AMf_20241120_125_samples_Eukaryome_Glomeromycota_genus.tree")
```