---
title: "Untitled"
output: html_document
date: "2025-05-07"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(raster)
library(mapview)
library(leaflet)
library(leafem)
library(rasterVis)
library(pals)
library(sf)
library(readxl)

library(Hmsc)
library(ape)
library(corrplot)
```

```{r}
metadata <- read_xlsx(path = "data/Metadata_Doreen_070525.xlsx")

utm_coords <- metadata %>%
  mutate(
    zone = as.numeric(substr(easting, 1, 2)),
    band = substr(easting, 3, 3),
    easting = as.numeric(sub(".* ", "", easting)),
    northing = as.numeric(northing),

    hemisphere = ifelse(band == "N", "north", "south") # M = South Hemisphere
  )

# Function to convert a single row to lat/lon
convert_to_wgs84 <- function(zone, hemisphere, easting, northing,Sample_code) {
  epsg <- if (hemisphere == "north") {
    32600 + zone  # Northern hemisphere UTM EPSG codes
  } else {
    32700 + zone  # Southern hemisphere UTM EPSG codes
  }

  point <- st_sfc(st_point(c(easting, northing)), crs = epsg)
  point_wgs84 <- st_transform(point, crs = 4326)
  coords <- st_coordinates(point_wgs84)
  tibble( lon = coords[1], lat = coords[2],Sample_code = Sample_code)
}
utm_coords %>%
  mutate(zone = case_when(zone==32~ 36,
                             .default = zone),
          northing = case_when(Sample_code == "N17" ~ 86562,
                             .default = northing)) -> utm_coords
# Apply to each row
wgs84_coords <- pmap_dfr(utm_coords[, c("zone", "hemisphere", "easting", "northing","Sample_code")], convert_to_wgs84) %>%
  left_join(metadata, by = "Sample_code") %>%
  st_as_sf(coords = c("lon", "lat"),crs = 4326) %>%
  mutate(
     lon= st_coordinates(.)[,1],
     lat= st_coordinates(.)[,2]
  )


leaflet() %>% 
  addTiles() %>%
  addMarkers(data=wgs84_coords,~lon, ~lat, popup = ~paste(round(lat,2), round(lon,2)) , label = ~Sample_code)%>%
  addCircleMarkers(data = wgs84_coords, lng = ~lon,
                  lat = ~lat, radius = 8, color = "blue", fillOpacity = 0.8)

leaflet() %>% addTiles() %>%
  addRasterImage(uganda_precip_2019, opacity = 0.3) 
```

```{r}
land_degradation_raster <- raster("data/Uganda_ldma_2020.tif/Uganda_ldma_2020.tif")
land_cover_raster <- raster("data/Uganda_Landcover_2014_Scheme_II/Uganda_Landcover_2014_Scheme_II.tif")
ndvi_uganda_lower_res <- raster("outputs/raster/ndvi_uganda_lower_res.tif")
uganda_precip_2019 <- raster("outputs/uganda_precip_2016_2021.tif")
plot(uganda_precip_2019)
str(land_degradation_raster)
land_degradation_raster <- ratify(land_degradation_raster)
levels(land_degradation_raster)[[1]]$land_degradation <- c("Very low", "Low", "Medium", "High", "Very High")
myPal1 <- rev(RColorBrewer::brewer.pal('PRGn', n=5))
myTheme1 <- rasterVis::rasterTheme(region = myPal1)

myPal2 <- pals::stepped2( n=18)
myTheme2 <- rasterVis::rasterTheme(region = myPal2)
levels(land_cover_raster)[[1]] %>%
  filter(Land_Cover != "No Data") -> levels(land_cover_raster)[[1]] 

rasterVis::levelplot(land_degradation_raster,par.settings = myTheme1)
rasterVis::levelplot(land_cover_raster,att ="Land_Cover", par.settings = myTheme2) 
plot(ndvi_uganda_lower_res)

plot(land_degradation_raster)
plot(as(wgs84_coords, "Spatial"), add = TRUE, col = "red")

land_cover_type <- levels(land_cover_raster)[[1]] %>%
  dplyr::select(ID,Land_Cover)
land_degradation_type <- levels(land_degradation_raster)[[1]]

land_degradation_df <- raster::extract(land_degradation_raster,  wgs84_coords)%>%
  tibble(ID = .,
    Sample_code  = wgs84_coords$Sample_code)%>%
  left_join(land_degradation_type,  by = "ID")

land_cover_df <- raster::extract(land_cover_raster, wgs84_coords)%>%
 tibble(ID = .,
    Sample_code  = wgs84_coords$Sample_code) %>%
    left_join(land_cover_type, by = "ID") 

ndvi_df <- raster::extract(ndvi_uganda_lower_res, wgs84_coords)%>%
 tibble(ndvi = .,
    Sample_code  = wgs84_coords$Sample_code)



wgs84_coords %>%
  left_join(land_cover_df, by = "Sample_code") %>%
  left_join(land_degradation_df, by = "Sample_code") %>%
  left_join(ndvi_df, by = "Sample_code") -> wgs84_coords

str(wgs84_coords)
pal_land_cover <- colorFactor(myPal2[unique(land_cover_df$ID)], 
                   domain = wgs84_coords$Land_Cover)

leaflet(wgs84_coords) %>%
  addTiles() %>%
  addCircleMarkers(
    ~lon, ~lat,
    color = "black",
    fillColor = ~pal_land_cover(Land_Cover),
    radius = 6,
    weight = 2,
    stroke = T,
    fillOpacity = 0.8,
    popup = ~Land_Cover,
    label = ~Sample_code,
    group = ~Land_Cover
  ) %>%
  addLegend(
    "bottomright",
    pal = pal_land_cover,
    values = ~Land_Cover,
    title = "Land Cover"
  )
 

pal_land_degradation <- colorFactor(myPal1[unique(land_degradation_df$land_degradation)], 
                   domain = wgs84_coords$land_degradation)

leaflet(wgs84_coords) %>%
  addTiles() %>%
  addCircleMarkers(
    ~lon, ~lat,
    color = "black",
    fillColor = ~pal_land_degradation(land_degradation),
    radius = 6,
    weight = 2,
    stroke = T,
    fillOpacity = 0.8,
    popup = ~land_degradation,
    label = ~Sample_code,
    group = ~land_degradation
  ) %>%
  addLegend(
    "bottomright",
    pal = pal_land_degradation,
    values = ~land_degradation,
    title = "Land Degradation"
  )
```



```{r}
data_18S <- "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2"
amf_taxonomic_levels <- c("kingdom", "domain", "phylum", "class", "order", "family",
                          "genus", "species")
amf_taxonomic_levels_eucaryome <- c("kingdom", "phylum", "class", "order", "family",
                          "genus", "species")
metadata <- as.data.frame(wgs84_coords) %>%
  dplyr::select(-geometry) %>%
  mutate(soil_code = paste("ROB_", Sample_code, "_s",sep=""))

amf_OTUs <- data_18S %>%
  #{{import data}}
  read_tsv() %>%
  #{{format column names}}
  rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~sub("_AMF_", "", .)) %>%
  rename_with(~sub("\\_AMF.*", "", .)) %>%
  rename_with(~sub("ROB_T_", "CONT_ROB_", .)) %>%
  rename_with(~sub("ROB_Tex", "CONT_ROB_T_ex", .)) %>%
  mutate(taxonomy = str_replace_all(taxonomy, ":","_"),
       taxonomy = str_replace_all(taxonomy, "[|]","; ")) %>%
  #{{select samples from ROBUST project}}
  dplyr::select(OTU, amplicon, taxonomy, starts_with("CONT", ignore.case = FALSE),
         starts_with("ROB", ignore.case = FALSE)) %>%
    #{{"Arbuscular mycorrhizal fungi & Mucoromycotina fine root endophytes" are selected}}
  #filter(str_detect(taxonomy,"Glomeromycotina")|str_detect(taxonomy,"Endogonales")) %>%
  filter(str_detect(taxonomy,"p__Glomeromycota")|str_detect(taxonomy,"o__Densosporales")) %>%
  mutate(taxonomy = str_replace_all(taxonomy, "(p__|c__|o__|f__|g__|s__)unclassified", "*"))%>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F) %>%
  #{{select only OTUs and controls}}
  droplevels()

taxonomy <- amf_OTUs %>%
  dplyr::select(amplicon, OTU, taxonomy) %>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F)
 


d <- amf_OTUs %>% 
  replace(amf_OTUs == 0, NA) %>%
  dplyr::select(amplicon, starts_with("CONT")) %>%
  pivot_longer(-amplicon, names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads))

amf_OTUs_decont <- amf_OTUs %>%
  replace(amf_OTUs == 0, NA) %>%
  pivot_longer(starts_with("ROB"), names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads)) %>%
  #{{merge with control samples}}
  left_join(count(d, amplicon, wt = reads), by = "amplicon") %>%
  #{{substract abundance of control samples}}
  mutate(reads = case_when(
    is.na(n)  ~ reads,
    n > reads ~ 0,
    TRUE      ~ reads - n)) %>%
  dplyr::select(-n,-starts_with("CONT")) %>%
  pivot_wider(names_from = samples, values_from =  reads, values_fill = 0) %>%
  as_tibble() %>%
  column_to_rownames(var = "amplicon") 
```


# Rarefaction
```{r}
plot.OTUcumSums <-function(x){plot(1:length(colSums(x)), log10(sort(colSums(x))+1))}
amf_OTUs_decont %>%
  dplyr::select(starts_with("ROB")) %>%
  plot.OTUcumSums
10^3.1
log10(10)
amf_OTUs_decont %>%
  dplyr::select(starts_with("ROB"))%>%
  colSums()%>%
  sort()


amf_OTUs_decont2 <- amf_OTUs_decont %>%
  #{{sample to exclude}}
  dplyr::select(starts_with("ROB"))%>%
  dplyr::select(where(~ sum(.x) >= 10^3))
smallest_amf <- min(colSums(amf_OTUs_decont2))

rarefaction.func <- function(df,rar_sample, n_rare = 100, rarcur = T){
  if(rarcur){ # {{ if rarefaction curve needed}}
    df %>%
      dplyr::select(starts_with("ROB")) %>%
      t() %>%
      vegan::rarecurve(step = 20, sample = rar_sample, col = "blue", cex = 0.6)
  }
  df %>%
    dplyr::select(starts_with("ROB")) %>%
    dim -> D
  # {{rarefaction}}
  df_rarefy = matrix(0, nrow =  D[1], ncol = D[2])
  for(n in 1:n_rare){
    df %>%
      dplyr::select(starts_with("ROB")) %>%
      t() %>%
      vegan::rrarefy(rar_sample) %>%
      t() + df_rarefy -> df_rarefy
    
  }
  round(df_rarefy/n_rare) -> df_rarefy
  df_rarefy %>%
    bind_cols(
      df %>%
        dplyr::select(!starts_with("ROB"))
    ) -> df_rarefy
  rownames(df_rarefy) = rownames(df)
  return(df_rarefy)
}
amf_OTUs_decont2 %>% 
  mutate(across(starts_with("ROB"), ~as.integer(.x))) -> amf_OTUs_decont2
set.seed(2121349)
rarefaction.func(df = amf_OTUs_decont2, rar_sample = smallest_amf, rarcur = T) -> amf_OTUs_rare
amf_OTUs_rare %>%
  as.data.frame()%>%
  rownames_to_column("amplicon") %>%
  left_join(amf_OTUs_decont %>%
              rownames_to_column("amplicon") %>%
              dplyr::select(-starts_with("ROB")), by ="amplicon") -> amf_OTUs_rare
```

# Abundance at family level

```{r}

amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, sep = "; ") %>%
  dplyr::select(taxonomy, starts_with("ROB")) %>%
  group_by(taxonomy)%>%
  mutate(
    family = paste(str_extract(
      taxonomy,"(f_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  dplyr::select(-taxonomy) %>%
  group_by(family) %>%
  summarise_all(sum) -> amf_family_rare
rownames(amf_family_rare)
  # if genus isn't known agglomerate a the lower taxonomic level known

amf_family_rare$family
```

```{r}
cluster_family <- amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, sep = "; ",remove = F) %>%
  dplyr::select(taxonomy, species, amplicon, OTU)

cluster_family%>% 
  filter(str_detect(taxonomy,"Planticonsort"))
```

# Algomerate phylogenetic tree at the family level

```{r}
tree_amf <- read.tree(file = "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples_Eukaryome_Glomeromycota.tree")

plot(tree_amf,show.tip.label = F)

```

```{r}
# 1. Compute full pairwise patristic distance matrix between OTUs
dist_mat <- cophenetic.phylo(tree_amf)  # matrix of tip-to-tip distances
min(dist_mat)
# 2. Prepare cluster mapping
otu_family <- cluster_family %>%
  group_by(taxonomy) %>%
  # if species isn't known agglomerate a the lower taxonomic level known
  group_by(taxonomy) %>%
  mutate(
    family = paste(str_extract(
      taxonomy,"(s_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  dplyr::select(-taxonomy) %>%
  filter(!(family %in% c("g_Rhizophagus_2", "g_Paraglomus_7")))


# 3. For each pair of species, compute median inter-cluster distance
family_list <- unique(otu_family$family)

# Create a matrix to store median distances
family_dist <- matrix(NA, nrow = length(family_list), ncol = length(family_list),
                       dimnames = list(family_list, family_list))

for (i in seq_along(family_list)) {
  for (j in seq_along(family_list)) {
    if (i <= j) {
      sp1_otus <- otu_family$amplicon[otu_family$family == family_list[i]]
      sp2_otus <- otu_family$amplicon[otu_family$family == family_list[j]]
      
      dists <- dist_mat[sp1_otus, sp2_otus, drop = FALSE]
      family_dist[i, j] <- family_dist[j, i] <- median(dists, na.rm = TRUE)
    }
  }
}

# 4. Create a new tree using median distances — convert matrix to phylo
family_tree <- nj(as.dist(family_dist))

# Optional: root or ladderize for plotting
family_tree <- ladderize(family_tree)

# 5. Plot the collapsed tree
plot(family_tree, cex = 0.7, main = "family-level Tree (collapsed from OTUs)")
```

# JSDM

```{r}

nChains = 12
test.run =F
if (test.run){
  #with this option, the vignette runs fast but results are not reliable
  thin = 5
  samples = 10
  transient = 5*thin
  verbose = 5*thin
} else {
  #with this option, the vignette evaluates slow but it reproduces the results of the
  #.pdf version
  thin = 5
  samples = 1000
  transient = 500*thin
  verbose = 500*thin
}
amf_family_rare %>%
  dplyr::select(-c(ROB_M_s, ROB_L_s)) %>%
  column_to_rownames("family") %>%
  filter(rowSums(across(starts_with("ROB")) != 0) > 3)%>%
  t() -> Y_family
colSums(Y_family) %>% sort

metadata %>% 
  filter(soil_code%in% rownames(Y_family))%>%
  mutate(across(where(is.character), ~as.factor(.x)),
         land_degradation = as.factor(land_degradation)) %>%
  dplyr::select(-Sample_code) %>%
  column_to_rownames("soil_code") -> metadata_filter

Y_family %>% 
  as.data.frame() %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  mutate(value = case_when(value == 0 ~ NA,
                             .default = value)) %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("soil_code") -> Y_family_trucated

Y_family_trucated %>% 
  as.data.frame() %>%
  pivot_longer(everything()) %>%
  ggplot() +
  facet_wrap(~name, scale = "free") +
  geom_histogram(aes(value))

family_tree_filter <- keep.tip(phy= family_tree, tip = colnames(Y_family_trucated))
dim(Y_family_trucated)
dim(metadata_filter)
str(family_tree_filter)
colnames(Y_family_trucated)

studyDesign = data.frame(sample = rownames(Y_family_trucated), stringsAsFactors=TRUE)
rL = HmscRandomLevel(units = studyDesign$sample)

m2 = Hmsc(Y = Y_family_trucated, XData = metadata_filter, 
          XFormula = ~Land_Cover + land_degradation + ndvi, 
          phyloTree =family_tree_filter,
          studyDesign = studyDesign,
          ranLevels = list(sample = rL),
          distr = "poisson")
load(file = "outputs/hmsc_family_ndvi.Rdata")
# m_converg2 = sampleMcmc(m2, thin = thin, samples = samples, transient = transient,
#                nChains = nChains, verbose = verbose, nParallel = 12)
# save(m_converg2, file = "outputs/hmsc_family_ndvi.Rdata")
```


```{r}
Y_family_trucated %>% 
  as.data.frame() %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  mutate(value = case_when(value != 0 ~ 1,
                             is.na(value) ~ 0)) %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("soil_code") ->  Y_family_bin
colnames(Y_family_bin)
metadata_filter$Land_Cover %>% unique
m2_bin = Hmsc(Y = Y_family_bin, XData = metadata_filter, 
          XFormula = ~ Land_Cover + land_degradation + ndvi , 
          phyloTree = family_tree_filter,
          studyDesign = studyDesign,
          ranLevels = list(sample = rL),
          distr = "probit")
load("outputs/hmsc_family_bin_ndvi.Rdata")

# m_converg2_bin = sampleMcmc(m2_bin, thin = thin, samples = samples, transient = transient,
#                nChains = nChains, verbose = verbose, nParallel = 12)
# save(m_converg2_bin, file = "outputs/hmsc_family_bin_ndvi.Rdata")
```

```{r}
mpost2 = convertToCodaObject(m_converg2)
gelman.diag(mpost2$Beta,multivariate=FALSE) -> gelman_diag_df2
gelman_diag_df2$psrf
postBeta2 = getPostEstimate(m_converg2, parName = "Beta") 

mpost_bin2 = convertToCodaObject(m_converg2_bin)
gelman.diag(mpost_bin2$Beta,multivariate=FALSE) -> gelman_diag_bin_df2
gelman_diag_bin_df2$psrf
postBeta_bin2 = getPostEstimate(m_converg2_bin, parName = "Beta") 

plotBeta(m_converg2, post = postBeta2, param = "Mean", plotTree = TRUE, supportLevel = 0.95, split=.5, spNamesNumbers = c(T,T))
plotBeta(m_converg2_bin, post = postBeta_bin2, param = "Mean", plotTree = TRUE, supportLevel = 0.95, split=.5, spNamesNumbers = c(T,T))
```

```{r}
library(nlme)
library(purrr)
models = list(abund = m_converg2,bin = m_converg2_bin)

summary(corPagel(0.5, phyloTree, fixed=F,form = ~names(y)))
compute.padgel.lambda<- function(models, i){
  
  m = models[["bin"]]
  postBeta = getPostEstimate(m, parName="Beta")
  phyloTree = m$phyloTree
  nc = dim(postBeta$mean)[1]
  results = matrix(NA,nrow=nrow(postBeta2$mean), ncol = 4) %>%
    as.data.frame() %>%
    rename(lambda = "V1", p_value = "V2", model = "V3", covariate = "V4") 
  control <- glsControl(opt = "optim", msVerbose = TRUE)
  for(j in 1:nc){
    y = postBeta$mean[j,]
    m1<-gls(y~1)
    m2<-gls(y~1,correlation=corPagel(0.5, phyloTree, fixed=F,form = ~names(y)),control = control)
    z1=summary(m2)
    z2=anova(m1,m2)
    lambda= z1$modelStruct[1]$corStruct
    pvalue=z2$`p-value`[2]
    results$p_value[j]=pvalue
    results$lambda[j]=lambda
    results$model[j]=i
    results$covariate[j]= colnames(m$X)[j]
  }
  return(results)
}
any(branching.times(phyloTree) == 0)
c("abund", "bin") %>%
  purrr::map(\(x) compute.padgel.lambda(models, x)) %>%
  list_rbind() -> lambda_df
lambda_df%>%
  filter(p_value<=0.05) %>%
  ggplot()+
  geom_point(aes(lambda,covariate, col = model), cex = 4)+
   theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", sixze=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))
  
```

```{r}
VP2 = computeVariancePartitioning(m_converg2,na.ignore =T) 
VP_bin2 = computeVariancePartitioning(m_converg2_bin,na.ignore= T) 

VP2$vals %>%
  as.data.frame %>%
  rownames_to_column("var") %>%
  pivot_longer(-var) %>%
  ggplot()+
  geom_col(aes(x= name, y= value, fill =var)) +
  theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", size=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))


VP_bin2$vals %>%
  as.data.frame %>%
  rownames_to_column("var") %>%
  pivot_longer(-var) %>%
  ggplot()+
  geom_col(aes(x= name, y= value, fill =var)) +
  theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", size=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))
```

```{r}
gelman.diag(mpost2$Rho)$psrf
summary(mpost2$Rho)

gelman.diag(mpost_bin2$Rho)$psrf
summary(mpost_bin2$Rho)
```

```{r}

OmegaCor = computeAssociations(m_converg2) 
supportLevel = 0.95 
toPlot = ((OmegaCor[[1]]$support>supportLevel) + (OmegaCor[[1]]$support<(1-supportLevel))>0)*OmegaCor[[1]]$mean 
corrplot(toPlot, method = "color", col=colorRampPalette(c("blue","white","red"))(200), tl.cex=.6, tl.col="black", title=paste("random effect level:", m_converg2$rLNames[1]), mar=c(0,0,1,0),order = "hclust")
```


Make raster same size and same pixels positions
```{r}
# land_degradation_lower_res <- raster::aggregate(land_degradation_raster, fact=20, fun = modal,  na.rm=TRUE)
# land_degradation_lower_res <- ratify(land_degradation_lower_res)
# levels(land_degradation_lower_res)[[1]]$land_degradation <- c("Very low", "Low", "Medium", "High", "Very High")
# 
# land_cover_resampled <- resample(land_cover_raster, land_degradation_lower_res, method = "ngb")
# land_cover_resampled <- ratify(land_cover_resampled)
# levels(land_cover_resampled)[[1]] <- levels(land_cover_raster)[[1]]
# raster::
# compareRaster(land_cover_resampled, land_degradation_lower_res)
# rasterVis::levelplot(land_cover_resampled,par.settings = myTheme2)
# rasterVis::levelplot(land_degradation_lower_res,par.settings = myTheme1)
# 
# writeRaster(land_degradation_lower_res,"outputs/raster/land_degradation_lower_res.tif",options=c('TFW=YES'))
# writeRaster(land_cover_resampled,"outputs/raster/land_cover_lower_res.tif",options=c('TFW=YES'))

land_degradation_lower_res <- raster("outputs/raster/land_degradation_lower_res.tif")
land_degradation_lower_res <- ratify(land_degradation_lower_res)
levels(land_degradation_lower_res)[[1]]$land_degradation <- c("Very low", "Low", "Medium", "High", "Very High")

land_cover_resampled <- raster("outputs/raster/land_cover_lower_res.tif")
land_cover_resampled <- ratify(land_cover_resampled)
levels(land_cover_resampled)[[1]] <- levels(land_cover_raster)[[1]]
```

Extract x value at each positions. Coffee plantation occur only at specific Land cover types thus I filter location (pixels) keep only those land cover types
```{r}
land_degradation_raster_df <- rasterToPoints(land_degradation_lower_res) %>%
  as.data.frame() %>%
  rename(land_degradation = "land_degradation_lower_res") %>%
  mutate(land_degradation = factor(land_degradation,labels = levels(land_degradation_lower_res)[[1]]$land_degradation))
  

land_cover_raster_df <- rasterToPoints(land_cover_resampled) %>%
  as.data.frame() %>%
  filter(land_cover_lower_res != 0) %>%
  left_join(land_cover_type, by = join_by(land_cover_lower_res == ID)) %>% 
  dplyr::select(-land_cover_lower_res)

ndvi_raster_df <- rasterToPoints(ndvi_uganda_lower_res,) %>%
  as.data.frame()%>%
  rename(ndvi = "ndvi_uganda_lower_res")

data_spatial_df <- land_cover_raster_df %>%
  left_join(land_degradation_raster_df, by = join_by("x", "y")) %>%
  left_join(ndvi_raster_df, by = join_by("x", "y")) %>%
  na.omit() %>%
  filter(Land_Cover %in% c(unique(metadata_filter$Land_Cover)))
```


```{r}
data_spatial_df$Land_Cover%>%unique
data_spatial_df %>%
  pivot_longer(c(Land_Cover,land_degradation)) %>%
  ggplot()+
  facet_wrap(~name)+
  geom_raster(aes(x,y,fill = value, group = name))
```


```{r}
library(viridis)
# data_spatial_df %>%
#   filter(x>32,x<33, y>0,y<2) -> test_df

data_spatial_df %>%
  pivot_longer(c(Land_Cover,land_degradation)) %>%
  ggplot()+
  facet_wrap(~name)+
  geom_raster(aes(x,y,fill = value))

data_spatial_df%>%
  dplyr::select(x,y) %>%
  as.matrix -> xy.grid

data_spatial_df%>%
  dplyr::select(-c(x,y)) -> XData.grid

Gradient <- prepareGradient(m2, XDataNew = XData.grid, sDataNew = list(route = xy.grid))
predY <- stats::predict(m_converg2, Gradient = Gradient)

Gradient2 <- prepareGradient(m2_bin, XDataNew = XData.grid, sDataNew = list(route = xy.grid))
predY2 <- stats::predict(m_converg2_bin, Gradient = Gradient2)

EpredY <- predY[[1]]
EpredY2 <- predY2[[1]]
for(i in 2 :length(predY)){
  EpredY <- EpredY + predY[[i]]
  if(i%%1000 == 0){
    print(i)
  }
}
EpredY <- EpredY/length(predY)
                        
for(i in 2 :1000){
  EpredY2 <- EpredY2 + predY2[[i]]
  if(i%%1000 == 0){
    print(i)
  }
}
EpredY2 <- EpredY2/1000

save(EpredY, file = "outputs/prediction_abudance_test1.Rdata")
save(EpredY2, file = "outputs/prediction_occurence_test1.Rdata")
# load(file = "outputs/prediction_abudance_test1.Rdata")
# load(file = file = "outputs/prediction_occurence_test1.Rdata")
data_spatial_df %>%
  bind_cols(EpredY) %>%
  pivot_longer(-c(x,y, Land_Cover, land_degradation, ndvi)) %>%
  mutate(value = log10(value+1)) %>%
  ggplot()+
  facet_wrap(~name)+
  geom_raster(aes(x,y,fill = value))+
  scale_fill_viridis(option = "cividis",direction = -1) +
  labs(x = " Lon", y = "Lat", fill = "Abundance (log10)") +
  theme_classic()+
  theme(aspect.ratio = 1,
        plot.margin = margin(0, 0, 0, 0),
        axis.title.x = element_blank(),
        panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        axis.title.y = element_text(colour = "black", size=14, face="italic"))
data_spatial_df %>%
  bind_cols(EpredY) %>%
  dplyr::select(-f__Densosporales_fam03_4)%>%
  pivot_longer(-c(x,y, Land_Cover, land_degradation, ndvi)) %>%
  mutate(value = log10(value+1)) %>%
  ggplot()+
  facet_wrap(~name)+
  geom_raster(aes(x,y,fill = value))+
  scale_fill_viridis(option = "cividis",direction = -1) +
  labs(x = " Lon", y = "Lat", fill = "Abundance (log10)") +
  theme_classic()+
  theme(aspect.ratio = 1,
        plot.margin = margin(0, 0, 0, 0),
        axis.title.x = element_blank(),
        panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        axis.title.y = element_text(colour = "black", size=14, face="italic"))

data_spatial_df %>%
  bind_cols(EpredY2) %>%
  pivot_longer(-c(x,y, Land_Cover, land_degradation, ndvi)) %>%
  ggplot()+
  facet_wrap(~name)+
  geom_raster(aes(x,y,fill = value))+
  scale_fill_viridis(option = "magma",direction = -1) +
  labs(x = " Lon", y = "Lat", fill = "Occurence probability") +
  theme_classic()+
  theme(aspect.ratio = 1,
        plot.margin = margin(0, 0, 0, 0),
        axis.title.x = element_blank(),
        panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        axis.title.y = element_text(colour = "black", size=14, face="italic"))

data_spatial_df %>%
  bind_cols(EpredY2) %>%
  pivot_longer(-c(x,y, Land_Cover, land_degradation)) %>%
  group_by(x,y) %>%
  summarise(richness = sum(value)) %>%
  ggplot()+
  geom_raster(aes(x,y,fill = richness))+
  scale_fill_viridis(option = "viridis",direction = -1) +
  labs(x = " Lon", y = "Lat", fill = "Family richness") +
  theme_classic()+
  theme(aspect.ratio = 1,
        plot.margin = margin(0, 0, 0, 0),
        axis.title.x = element_blank(),
        panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        axis.title.y = element_text(colour = "black", size=14, face="italic"))
save.image()
```

