---
title: "Untitled"
output: html_document
date: "2025-04-18"
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
library(tidyverse)
library(vegan)
library(ggsci)
library(ape)
library(ggforce)
# install.packages("BiocManager")
# BiocManager::install("Biostrings")
# devtools::install_github('reptalex/phylofactor')
library(phylofactor)
library(ape)
library(alluvial)
library(aricode)
library(GGally)
library(sbm)
library(readxl)
library(GGally)
library(FactoMineR)
library(factoextra)
library(Hmsc)
```


```{r}
data_18S <- "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2"
amf_taxonomic_levels <- c("kingdom", "domain", "phylum", "class", "order", "family",
                          "genus", "species")
amf_taxonomic_levels_eucaryome <- c("kingdom", "phylum", "class", "order", "family",
                          "genus", "species")
metadata <- read_tsv("data/metadata.txt")

amf_OTUs <- data_18S %>%
  #{{import data}}
  read_tsv() %>%
  #{{format column names}}
  rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~sub("_AMF_", "", .)) %>%
  rename_with(~sub("\\_AMF.*", "", .)) %>%
  rename_with(~sub("ROB_T_", "CONT_ROB_", .)) %>%
  rename_with(~sub("ROB_Tex", "CONT_ROB_T_ex", .)) %>%
  mutate(taxonomy = str_replace_all(taxonomy, ":","_"),
       taxonomy = str_replace_all(taxonomy, "[|]","; ")) %>%
  #{{select samples from ROBUST project}}
  select(OTU, amplicon, taxonomy, starts_with("CONT", ignore.case = FALSE),
         starts_with("ROB", ignore.case = FALSE)) %>%
    #{{"Arbuscular mycorrhizal fungi & Mucoromycotina fine root endophytes" are selected}}
  #filter(str_detect(taxonomy,"Glomeromycotina")|str_detect(taxonomy,"Endogonales")) %>%
  filter(str_detect(taxonomy,"p__Glomeromycota")|str_detect(taxonomy,"o__Densosporales")) %>%
  mutate(taxonomy = str_replace_all(taxonomy, "(p__|c__|o__|f__|g__|s__)unclassified", "*"))%>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F) %>%
  #{{select only OTUs and controls}}
  droplevels()

taxonomy <- amf_OTUs %>%
  select(amplicon, OTU, taxonomy) %>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F)
 


d <- amf_OTUs %>% 
  replace(amf_OTUs == 0, NA) %>%
  select(amplicon, starts_with("CONT")) %>%
  pivot_longer(-amplicon, names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads))

amf_OTUs_decont <- amf_OTUs %>%
  replace(amf_OTUs == 0, NA) %>%
  pivot_longer(starts_with("ROB"), names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads)) %>%
  #{{merge with control samples}}
  left_join(count(d, amplicon, wt = reads), by = "amplicon") %>%
  #{{substract abundance of control samples}}
  mutate(reads = case_when(
    is.na(n)  ~ reads,
    n > reads ~ 0,
    TRUE      ~ reads - n)) %>%
  select(-n,-starts_with("CONT")) %>%
  pivot_wider(names_from = samples, values_from =  reads, values_fill = 0) %>%
  as_tibble() %>%
  column_to_rownames(var = "amplicon") 
```


# Rarefaction
```{r}
plot.OTUcumSums <-function(x){plot(1:length(colSums(x)), log10(sort(colSums(x))+1))}
amf_OTUs_decont %>%
  select(starts_with("ROB")) %>%
  plot.OTUcumSums
10^3.1
log10(10)
amf_OTUs_decont %>%
  select(starts_with("ROB"))%>%
  colSums()%>%
  sort()


amf_OTUs_decont2 <- amf_OTUs_decont %>%
  #{{sample to exclude}}
  select(starts_with("ROB"))%>%
  select(where(~ sum(.x) >= 10^3))
smallest_amf <- min(colSums(amf_OTUs_decont2))

rarefaction.func <- function(df,rar_sample, n_rare = 100, rarcur = T){
  if(rarcur){ # {{ if rarefaction curve needed}}
    df %>%
      select(starts_with("ROB")) %>%
      t() %>%
      vegan::rarecurve(step = 20, sample = rar_sample, col = "blue", cex = 0.6)
  }
  df %>%
    select(starts_with("ROB")) %>%
    dim -> D
  # {{rarefaction}}
  df_rarefy = matrix(0, nrow =  D[1], ncol = D[2])
  for(n in 1:n_rare){
    df %>%
      select(starts_with("ROB")) %>%
      t() %>%
      vegan::rrarefy(rar_sample) %>%
      t() + df_rarefy -> df_rarefy
    
  }
  round(df_rarefy/n_rare) -> df_rarefy
  df_rarefy %>%
    bind_cols(
      df %>%
        select(!starts_with("ROB"))
    ) -> df_rarefy
  rownames(df_rarefy) = rownames(df)
  return(df_rarefy)
}
amf_OTUs_decont2 %>% 
  mutate(across(starts_with("ROB"), ~as.integer(.x))) -> amf_OTUs_decont2
set.seed(2121349)
rarefaction.func(df = amf_OTUs_decont2, rar_sample = smallest_amf, rarcur = T) -> amf_OTUs_rare
amf_OTUs_rare %>%
  as.data.frame()%>%
  rownames_to_column("amplicon") %>%
  left_join(amf_OTUs_decont %>%
              rownames_to_column("amplicon") %>%
              select(-starts_with("ROB")), by ="amplicon") -> amf_OTUs_rare
```
# Abundance at specie level

```{r}
amf_OTUs_rare %>% 
  column_to_rownames("amplicon") %>%
  select(starts_with("ROB")) -> amf_OTUs_rare_matrix
amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, genus, sep = "; ") %>%
  select(taxonomy, starts_with("ROB")) %>%
  group_by(taxonomy)%>%
  mutate(
    genus = paste(str_extract(
      taxonomy,"(g_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  select(-taxonomy) %>%
  group_by(genus) %>%
  summarise_all(sum) -> amf_genus_rare
  # if genus isn't known agglomerate a the lower taxonomic level known
amf_OTUs_rare$taxonomy
amf_genus_rare$genus
```

```{r}
# Step 6: For each cluster, assign a species name if one exists

cluster_genus <- amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family,genus, sep = "; ",remove = F) %>%
  select(taxonomy, amplicon, species, amplicon, OTU)
```

```{r}
tree_amf <- read.tree(file = "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples_Eukaryome_Glomeromycota.tree")
pf.heatmap(tree=tree_amf,Data=log1p(amf_OTUs_decont2),color = NA)
pf.heatmap(tree=tree_amf,Data=log1p(amf_OTUs_rare_matrix))
rownames(amf_OTUs_rare_matrix) %in% tree_amf$tip.label
tree_amf$tip.label %in% rownames(amf_OTUs_rare_matrix)
data.frame(amplicon = rownames(amf_OTUs_rare_matrix)) %>%
  write.table( "amplicon_afm.txt")

plot(tree_amf,show.tip.label = F)

```

# Algomerate phylogenetic tree at the species level

```{r}
# 1. Compute full pairwise patristic distance matrix between OTUs
dist_mat <- cophenetic.phylo(tree_amf)  # matrix of tip-to-tip distances

# 2. Prepare cluster mapping
otu_genus <- cluster_genus %>%
  # if species isn't known agglomerate a the lower taxonomic level known
  group_by(taxonomy) %>%
  mutate(
    genus = paste(str_extract(
      taxonomy,"(g_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  select(-taxonomy)


# 3. For each pair of species, compute median inter-cluster distance
genus_list <- unique(otu_genus$genus)

# Create a matrix to store median distances
genus_dist <- matrix(NA, nrow = length(genus_list), ncol = length(genus_list),
                       dimnames = list(genus_list, genus_list))

for (i in seq_along(genus_list)) {
  for (j in seq_along(genus_list)) {
    if (i <= j) {
      sp1_otus <- otu_genus$amplicon[otu_genus$genus == genus_list[i]]
      sp2_otus <- otu_genus$amplicon[otu_genus$genus == genus_list[j]]
      
      dists <- dist_mat[sp1_otus, sp2_otus, drop = FALSE]
      genus_dist[i, j] <- genus_dist[j, i] <- median(dists, na.rm = TRUE)
    }
  }
}

# 4. Create a new tree using median distances — convert matrix to phylo
genus_tree <- nj(as.dist(genus_dist))

# Optional: root or ladderize for plotting
genus_tree <- ladderize(genus_tree)

# 5. Plot the collapsed tree
plot(genus_tree, cex = 0.7, main = "Genus-level Tree (collapsed from OTUs)")
```

# JSDM


```{r}

soil_df <- read_xlsx("data/soil_data_dry_season_2023.xlsx") %>%
  rename(coffee_system = "Robusta coffee system", soil_code = "SAMPLE NO", sand = "Sand%", clay = "Clay%", silt = "Silt%", pH ="soil pH", 
         OC = "OC%", MO = "OM%", N = "N%", P = "P/mg/kg", K = "K/Cmol/kg", Na = "Na/Cmol/kg") %>%
  mutate(soil_code = str_c("ROB_", soil_code,"_s"))

soil_df %>%
  select(-c(soil_code, Replicate,Zone, Cluster, coffee_system, Replicate)) %>%
 
  ggpairs()

soil_df %>%
  select(-c(soil_code, Replicate,Zone, Cluster, coffee_system, Replicate, OC,silt, clay)) %>%
  decostand(method = "standardize")%>%
  PCA() -> PCA_soil
str(soil_df)
fviz_pca_biplot(PCA_soil,col.ind = soil_df$Cluster)

soil_df %>%
  select(N,P,K) %>%
  decostand(method = "standardize")%>%
  PCA() -> PCA_npk

fviz_pca_biplot(PCA_npk,col.ind = soil_df$Cluster)

HCPC_npk =PCA_npk%>%
  HCPC(-1)

HCPC_npk$data.clust%>%
  rownames_to_column("id") %>%
  arrange(clust) -> id_pca_npk_clust

id_pca_npk_clust[id_pca_npk_clust$id, 2:4]%>%
  as.matrix()%>%
  t() %>%
  plotMyMatrix(dimLabels = c("Soil", "Grid"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(id_pca_npk_clust$clust))[-1]-0.5, col ="red")+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))
```
```{r}
amf_genus_rare %>%
  select(-c(ROB_M_s, ROB_L_s)) %>%
  column_to_rownames("genus") %>%
  filter(rowSums(across(starts_with("ROB")) != 0) > 3)%>%
  t()-> Y
```

```{r}
zscore_normalization <- function(data) {
  # Step 1: Rank the data
  ranked_data <- rank(data)
  n <- length(data)
  
  # Step 2: Calculate quantile positions
  quantile_positions <- (ranked_data) / (n + 1)
  
  # Step 3: Transform quantiles to z-scores
  mean_std_normal <- qnorm(quantile_positions)
  
  return(mean_std_normal)
}
logit <- function(p){
  p<-p/100
  log(p/(1-p))}

soil_df %>%
  dplyr::select(sand, clay, silt, pH, MO, N, P, K, Na,soil_code) %>%
  filter(soil_code%in% rownames(Y )) %>%
  column_to_rownames("soil_code") -> soil_df2
ggpairs(soil_df2)
soil_df2 %>%
  mutate(across(c(P, K, Na),  ~log(.x)),
         
         across(c(sand, clay, silt),~logit(.x)),
         across(c( pH, MO, N, P, K, Na, sand, clay, silt),  ~decostand(.x, method = "standardize"))) %>%
  t()-> soil_df2_standa

ggpairs(t(soil_df2_standa))

soil_df2_standa%>%
  as.matrix() %>%
  estimateBipartiteSBM(
    model = 'gaussian', 
    dimLabels = c(row = "var", col = "soil_code")) -> smb_soil

sbm_soil = as.factor(smb_soil$memberships$soil_code) ##### Grids cluster for soil

code_id = smb_soil$memberships[2]%>%
  as.data.frame()%>%
  rownames_to_column("id")%>%
  mutate(id = as.numeric(id))%>%
  arrange(soil_code)

var_id = smb_soil$memberships[1]%>%
  as.data.frame()%>%
  rownames_to_column("id")%>%
  mutate(id = as.numeric(id))%>%
  arrange(var)


soil_df2_standa[var_id$id,code_id$id]%>%
  as.matrix()%>%
  plotMyMatrix(dimLabels = c("Soil", "Grid"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(code_id$soil_code))[-1]-0.5, col ="red")+
  geom_hline(yintercept = abs(which(!duplicated(var_id$var))[-1]-nrow(var_id)-1)+0.5, col = "red", linetype = 5)+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))

t(decostand(soil_df2,method = "standardize"))[var_id$id,code_id$id]%>%
  as.matrix()%>%
  plotMyMatrix(dimLabels = c("Soil", "Grid"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(code_id$soil_code))[-1]-0.5, col ="red")+
  geom_hline(yintercept = abs(which(!duplicated(var_id$var))[-1]-nrow(var_id)-1)+0.5, col = "red", linetype = 5)+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))

culster_soil_df <- data.frame(soil_code = colnames(soil_df2_standa), culster_soil = as.factor(smb_soil$memberships$soil_code))

```

```{r}
soil_df2%>%
  select(N,P,K) %>%
  mutate(across(everything(),~zscore_normalization(.x))) %>%
  as.matrix() %>%
  t() %>%
  estimateBipartiteSBM(
    model = 'gaussian', 
    dimLabels = c(row = "var", col = "soil_code")) -> smb_NPK
plot(smb_NPK)
```

```{r}
soil_df %>%
  left_join(culster_soil_df, by ="soil_code") -> soil_df
```


```{r}
nChains = 12
test.run = F
if (test.run){
  #with this option, the vignette runs fast but results are not reliable
  thin = 1
  samples = 10
  transient = 5
  verbose = 5
} else {
  #with this option, the vignette evaluates slow but it reproduces the results of the
  #.pdf version
  thin = 10
  samples = 1000
  transient = 500*thin
  verbose = 500*thin
}

soil_df %>% 
  filter(soil_code%in% rownames(Y))%>%
  mutate(across(where(is.character), ~as.factor(.x)),
         coffee_system = factor(coffee_system,levels = c("RCM","RCB", "RCT", "RCTB", "RCTBA"))) %>%
  column_to_rownames("soil_code") -> soil_df3

Y%>%
  as.data.frame() %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  mutate(value = case_when(value == 0 ~ NA,
                             .default = value)) %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("soil_code") -> Y_trucated
 
Y_trucated %>% 
  as.data.frame() %>%
  pivot_longer(everything()) %>%
  ggplot() +
  facet_wrap(~name, scale = "free") +
  geom_histogram(aes(value))

genus_tree_filter <- keep.tip(phy= genus_tree, tip = colnames(Y_trucated))
dim(Y_trucated)
dim(soil_df3)
str(genus_tree_filter)

studyDesign = data.frame(sample = rownames(Y), stringsAsFactors=TRUE)
rL = HmscRandomLevel(units = studyDesign$sample)


m1 = Hmsc(Y = Y_trucated, XData = soil_df3, 
          XFormula = ~coffee_system + Cluster + culster_soil, 
          phyloTree =genus_tree_filter,
          studyDesign = studyDesign,
          ranLevels = list(sample = rL),
          distr = "poisson")


# m_converg = sampleMcmc(m1, thin = thin, samples = samples, transient = transient,
#                nChains = nChains, verbose = verbose, nParallel = 12)
# save(m_converg,file = "outputs/hmsc_genus.Rdata")
load("outputs/hmsc_genus.Rdata")
```


```{r}
Y_trucated %>% 
  as.data.frame() %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  mutate(value = case_when(value != 0 ~ 1,
                             is.na(value) ~ 0)) %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("soil_code")-> Y_bin

m1_bin = Hmsc(Y = Y_bin, XData = soil_df3, 
          XFormula = ~coffee_system + Cluster + culster_soil, 
          phyloTree = genus_tree_filter,
          studyDesign = studyDesign,
          ranLevels = list(sample = rL),
          distr = "probit")
# m_converg_bin = sampleMcmc(m1_bin, thin = thin, samples = samples, transient = transient,
#                nChains = nChains, verbose = verbose, nParallel = 12)
# save(m_converg_bin,file = "outputs/hmsc_genus_bin.Rdata")
load("outputs/hmsc_genus_bin.Rdata")
```

```{r}
Y %>%
  as.data.frame %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  
  left_join(soil_df3 %>% 
              rownames_to_column("soil_code"), by = "soil_code") -> sp_env_df

sp_env_df %>%
  ggplot()+
  facet_wrap(~name, scale = "free") +
  geom_boxplot(aes(coffee_system, value))
sp_env_df %>%
  ggplot()+
  facet_wrap(~name, scale = "free") +
  geom_boxplot(aes(culster_soil, value))

sp_env_df %>%
  ggplot()+
  facet_wrap(~name, scale = "free") +
  geom_boxplot(aes(Cluster, value))
```

```{r}

mpost = convertToCodaObject(m_converg)
gelman.diag(mpost$Beta,multivariate=FALSE) -> gelman_diag_df
gelman_diag_df$psrf
postBeta = getPostEstimate(m_converg, parName = "Beta") 

mpost_bin = convertToCodaObject(m_converg_bin)
gelman.diag(mpost_bin$Beta,multivariate=FALSE) -> gelman_diag_bin_df
gelman_diag_bin_df$psrf
postBeta_bin = getPostEstimate(m_converg_bin, parName = "Beta") 

plotBeta(m_converg, post = postBeta, param = "Mean", plotTree = TRUE, supportLevel = 0.95, split=.5, spNamesNumbers = c(T,T))
plotBeta(m_converg_bin, post = postBeta_bin, param = "Mean", plotTree = TRUE, supportLevel = 0.95, split=.5, spNamesNumbers = c(T,T))
```
```{r}
soil_df3 %>%
  ggplot() +
  geom_bin2d(aes(x = Cluster,y = coffee_system))

soil_df3 %>%
  ggplot() +
  geom_bin2d(aes(x = Cluster,culster_soil))
soil_df3 %>%
  ggplot() +
  geom_bin2d(aes(x = coffee_system,culster_soil))

```

```{r}

VP = computeVariancePartitioning(m_converg,na.ignore =T) 
VP_bin = computeVariancePartitioning(m_converg_bin,na.ignore= T) 

VP$vals %>%
  as.data.frame %>%
  rownames_to_column("var") %>%
  pivot_longer(-var) %>%
  ggplot()+
  geom_col(aes(x= name, y= value, fill =var)) +
  theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", size=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))


VP_bin$vals %>%
  as.data.frame %>%
  rownames_to_column("var") %>%
  pivot_longer(-var) %>%
  ggplot()+
  geom_col(aes(x= name, y= value, fill =var)) +
  theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", size=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))
```

```{r}
gelman.diag(mpost$Rho)$psrf
summary(mpost$Rho)

gelman.diag(mpost_bin$Rho)$psrf
summary(mpost_bin$Rho)
```

```{r}
site_loading <- getPostEstimate(m_converg, "Eta")$mean %>%
  as.data.frame() %>%
  rename_with(~paste("axes",1:8, sep="")) %>%
  mutate(soil_code = rownames(Y_family)) %>%
  left_join(soil_df3 %>%
              rownames_to_column("soil_code"), by = "soil_code")

species_loading <- getPostEstimate(m_converg, "Lambda")$mean %>%
  as.data.frame() %>%
  rename_with(~colnames(Y_trucated)) %>%
  mutate(axes = paste("axes",1:8, sep="")) %>%
  column_to_rownames("axes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("taxa") 

ggplot()+
  geom_point(data= site_loading, aes(axes1,axes2, fill = Cluster),shape =21, size = 2)+
  stat_ellipse(data= site_loading, aes(axes1,axes2, col = Cluster),type ="euclid")+
  geom_point(data= species_loading, aes(axes1,axes2), shape =22, fill = "black", size = 2) +
  geom_text(data= species_loading, aes(axes1,axes2,label = taxa), size = 2) +
  theme_bw()

site_loading_bin <- getPostEstimate(m_converg_bin, "Eta")$mean %>%
  as.data.frame() %>%
  rename_with(~paste("axes",1:5, sep="")) %>%
  mutate(soil_code = rownames(Y_trucated)) %>%
  left_join(soil_df3 %>%
              rownames_to_column("soil_code"), by = "soil_code")

species_loading_bin <- getPostEstimate(m_converg_bin, "Lambda")$mean %>%
  as.data.frame() %>%
  rename_with(~colnames(Y_trucated)) %>%
  mutate(axes = paste("axes",1:5, sep="")) %>%
  column_to_rownames("axes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("taxa") 

ggplot()+
  geom_point(data= site_loading_bin, aes(axes1,axes2, fill = Cluster),shape =21, size = 2)+
  stat_ellipse(data= site_loading_bin, aes(axes1,axes2, col = Cluster),type ="euclid")+
  geom_point(data= species_loading_bin, aes(axes1,axes2), shape =22, fill = "black", size = 2) +
  geom_text(data= species_loading_bin, aes(axes1,axes2,label = taxa), size = 2) +
  theme_bw()
```

```{r}
Gradient = constructGradient(m_converg,focalVariable ="coffee_system")
predY = predict(m_converg, XData=Gradient$XDataNew, studyDesign=Gradient$studyDesignNew, ranLevels=Gradient$rLNew, expected=TRUE) 
plotGradient(m_converg, Gradient, pred=predY, measure="Y",index = 2, showData = TRUE, jigger = 0.2)

```


# Abundance at family level

```{r}

amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, sep = "; ") %>%
  select(taxonomy, starts_with("ROB")) %>%
  group_by(taxonomy)%>%
  mutate(
    family = paste(str_extract(
      taxonomy,"(f_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  select(-taxonomy) %>%
  group_by(family) %>%
  summarise_all(sum) -> amf_family_rare
rownames(amf_family_rare)
  # if genus isn't known agglomerate a the lower taxonomic level known

amf_family_rare$family
```

```{r}
cluster_family <- amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, sep = "; ",remove = F) %>%
  select(taxonomy, species, amplicon, OTU)
```

# Algomerate phylogenetic tree at the family level

```{r}
# 1. Compute full pairwise patristic distance matrix between OTUs
dist_mat <- cophenetic.phylo(tree_amf)  # matrix of tip-to-tip distances

# 2. Prepare cluster mapping
otu_family <- cluster_family %>%
  group_by(taxonomy) %>%
  # if species isn't known agglomerate a the lower taxonomic level known
  group_by(taxonomy) %>%
  mutate(
    family = paste(str_extract(
      taxonomy,"(s_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  select(-taxonomy) %>%
  filter(!(family %in% c("g_Rhizophagus_2", "g_Paraglomus_7")))


# 3. For each pair of species, compute median inter-cluster distance
family_list <- unique(otu_family$family)

# Create a matrix to store median distances
family_dist <- matrix(NA, nrow = length(family_list), ncol = length(family_list),
                       dimnames = list(family_list, family_list))

for (i in seq_along(family_list)) {
  for (j in seq_along(family_list)) {
    if (i <= j) {
      sp1_otus <- otu_family$amplicon[otu_family$family == family_list[i]]
      sp2_otus <- otu_family$amplicon[otu_family$family == family_list[j]]
      
      dists <- dist_mat[sp1_otus, sp2_otus, drop = FALSE]
      family_dist[i, j] <- family_dist[j, i] <- median(dists, na.rm = TRUE)
    }
  }
}

# 4. Create a new tree using median distances — convert matrix to phylo
family_tree <- nj(as.dist(family_dist))

# Optional: root or ladderize for plotting
family_tree <- ladderize(family_tree)

# 5. Plot the collapsed tree
plot(family_tree, cex = 0.7, main = "family-level Tree (collapsed from OTUs)")
```

# JSDM

```{r}
library(Hmsc)

nChains = 12
test.run =F
if (test.run){
  #with this option, the vignette runs fast but results are not reliable
  thin = 5
  samples = 10
  transient = 5*thin
  verbose = 5*thin
} else {
  #with this option, the vignette evaluates slow but it reproduces the results of the
  #.pdf version
  thin = 5
  samples = 1000
  transient = 500*thin
  verbose = 500*thin
}
amf_family_rare %>%
  select(-c(ROB_M_s, ROB_L_s)) %>%
  column_to_rownames("family") %>%
  filter(rowSums(across(starts_with("ROB")) != 0) > 3)%>%
  t() -> Y_family
colSums(Y_family) %>% sort


Y_family %>% 
  as.data.frame() %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  mutate(value = case_when(value == 0 ~ NA,
                             .default = value)) %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("soil_code") -> Y_family_trucated

Y_family_trucated %>% 
  as.data.frame() %>%
  pivot_longer(everything()) %>%
  ggplot() +
  facet_wrap(~name, scale = "free") +
  geom_histogram(aes(value))

family_tree_filter <- keep.tip(phy= family_tree, tip = colnames(Y_family_trucated))
dim(Y_family_trucated)
dim(soil_df3)
str(family_tree_filter)
colnames(Y_family_trucated)

studyDesign = data.frame(sample = rownames(Y_family_trucated), stringsAsFactors=TRUE)
rL = HmscRandomLevel(units = studyDesign$sample)

m2 = Hmsc(Y = Y_family_trucated, XData = soil_df3, 
          XFormula = ~coffee_system + Cluster + culster_soil, 
          phyloTree =family_tree_filter,
          studyDesign = studyDesign,
          ranLevels = list(sample = rL),
          distr = "poisson")
load(file = "outputs/hmsc_family.Rdata")
# m_converg2 = sampleMcmc(m2, thin = thin, samples = samples, transient = transient,
#                nChains = nChains, verbose = verbose, nParallel = 12)
# save(m_converg2, file = "outputs/hmsc_family.Rdata")
```


```{r}
Y_family_trucated %>% 
  as.data.frame() %>%
  rownames_to_column("soil_code") %>%
  pivot_longer(-soil_code) %>%
  mutate(value = case_when(value != 0 ~ 1,
                             is.na(value) ~ 0)) %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("soil_code") ->  Y_family_bin
colnames(Y_family_bin)
m2_bin = Hmsc(Y = Y_family_bin, XData = soil_df3, 
          XFormula = ~coffee_system + Cluster + culster_soil, 
          phyloTree = family_tree_filter,
          studyDesign = studyDesign,
          ranLevels = list(sample = rL),
          distr = "probit")

load(file = "outputs/hmsc_family_bin.Rdata")
# m_converg2_bin = sampleMcmc(m2_bin, thin = thin, samples = samples, transient = transient,
#                nChains = nChains, verbose = verbose, nParallel = 12)
# save(m_converg2_bin, file = "outputs/hmsc_family_bin.Rdata")
```

```{r}

mpost2 = convertToCodaObject(m_converg2)
gelman.diag(mpost2$Beta,multivariate=FALSE) -> gelman_diag_df2
gelman_diag_df2$psrf
postBeta2 = getPostEstimate(m_converg2, parName = "Beta") 

mpost_bin2 = convertToCodaObject(m_converg2_bin)
gelman.diag(mpost_bin2$Beta,multivariate=FALSE) -> gelman_diag_bin_df2
gelman_diag_bin_df2$psrf
postBeta_bin2 = getPostEstimate(m_converg2_bin, parName = "Beta") 

plotBeta(m_converg2, post = postBeta2, param = "Mean", plotTree = TRUE, supportLevel = 0.95, split=.5, spNamesNumbers = c(T,T))
plotBeta(m_converg2_bin, post = postBeta_bin2, param = "Mean", plotTree = TRUE, supportLevel = 0.95, split=.5, spNamesNumbers = c(T,T))
```

```{r}

VP2 = computeVariancePartitioning(m_converg2,na.ignore =T) 
VP_bin2 = computeVariancePartitioning(m_converg2_bin,na.ignore= T) 

VP2$vals %>%
  as.data.frame %>%
  rownames_to_column("var") %>%
  pivot_longer(-var) %>%
  ggplot()+
  geom_col(aes(x= name, y= value, fill =var)) +
  theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", size=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))


VP_bin2$vals %>%
  as.data.frame %>%
  rownames_to_column("var") %>%
  pivot_longer(-var) %>%
  ggplot()+
  geom_col(aes(x= name, y= value, fill =var)) +
  theme_classic()+
    theme(
          axis.title.y = element_blank(),
          axis.text.x = element_text(colour = "black", size=13, face="italic", angle = 90, vjust =0, hjust =1),
          axis.text.y = element_text(colour = "black", size=13, face="italic"),
          axis.title = element_text(colour = "black", size=16),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
          legend.title = element_text(colour = "black", size=16,
                                      hjust =0.5),
  
          legend.text = element_text(colour = "black", size=13))
```

```{r}
gelman.diag(mpost2$Rho)$psrf
summary(mpost2$Rho)

gelman.diag(mpost_bin2$Rho)$psrf
summary(mpost_bin2$Rho)
```

```{r}
library(corrplot)
OmegaCor = computeAssociations(m_converg2) 
supportLevel = 0.95 
toPlot = ((OmegaCor[[1]]$support>supportLevel) + (OmegaCor[[1]]$support<(1-supportLevel))>0)*OmegaCor[[1]]$mean 
corrplot(toPlot, method = "color", col=colorRampPalette(c("blue","white","red"))(200), tl.cex=.6, tl.col="black", title=paste("random effect level:", m_converg2$rLNames[1]), mar=c(0,0,1,0),order = "hclust")
```

```{r}
gelman.diag(mpost2$Rho)$psrf
summary(mpost2$Rho)
```

```{r}
site_loading <- getPostEstimate(m_converg2, "Eta")$mean %>%
  as.data.frame() %>%
  rename_with(~paste("axes",1:5, sep="")) %>%
  mutate(soil_code = rownames(Y_family)) %>%
  left_join(soil_df3 %>%
              rownames_to_column("soil_code"), by = "soil_code")

species_loading <- getPostEstimate(m_converg2, "Lambda")$mean%>%
  as.data.frame() %>%
  rename_with(~colnames(Y_family)) %>%
  mutate(axes = paste("axes",1:5, sep="")) %>%
  column_to_rownames("axes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("taxa") 

ggplot()+
  geom_point(data= site_loading, aes(axes1,axes2, fill = Cluster),shape =21, size = 2)+
  stat_ellipse(data= site_loading, aes(axes1,axes2, col = Cluster),type ="euclid")+
  geom_point(data= species_loading, aes(axes1,axes2), shape =22, fill = "black", size = 2) +
  geom_text(data= species_loading, aes(axes1,axes2,label = taxa), size = 2) +
  theme_bw()
  

par(mfrow = c(2,2))
biPlot(m_converg2, etaPost = getPostEstimate(m_converg2, "Eta"), lambdaPost = getPostEstimate(m_converg2, "Lambda"),colVar = 1)
biPlot(m_converg2, etaPost = getPostEstimate(m_converg2, "Eta"), lambdaPost = getPostEstimate(m_converg2, "Lambda"),colVar = 2)
biPlot(m_converg2, etaPost = getPostEstimate(m_converg2, "Eta"), lambdaPost = getPostEstimate(m_converg2, "Lambda"),colVar = 3)
```

```{r}
Gradient2 = constructGradient(m_converg2, focalVariable ="MO")
predY2 = predict(m_converg2, XData=Gradient2$XDataNew, studyDesign=Gradient2$studyDesignNew, ranLevels=Gradient2$rLNew, expected=TRUE) 
par(mfrow = c(4,4))
for(i in 1:14){
plotGradient(m_converg2, Gradient2, pred=predY2, measure="Y",index = i, showData = TRUE)
}

# Gradient3 = constructGradient(m_converg2, focalVariable ="coffee_system")
# predY3 = predict(m_converg2, XData=Gradient3$XDataNew, studyDesign=Gradient3$studyDesignNew, ranLevels=Gradient3$rLNew, expected=TRUE) 
# par(mfrow = c(4,4))
# for(i in 1:14){
# plotGradient(m_converg2, Gradient3, pred=predY3, measure="Y",index = i, showData = TRUE) 
#   }

```

## sbm

```{r}
amf_OTUs_rare %>%
  select(starts_with("ROB"), genus,amplicon )%>%
  select(-c(ROB_M_s, ROB_L_s)) %>%
  filter(genus=="g_Glomus")%>%
  select(-genus) %>%
  column_to_rownames("amplicon") %>%
  as.matrix() %>%
  t() -> temp
colSums(temp) %>% sort %>% names-> sort_abundance
temp[,sort_abundance]%>%
  log1p()%>%
  plotMyMatrix()
hist(log1p(temp))
Y_family -> Y_family_na 
Y_family_na[Y_family_na==0] <-NA 
t(Y_family!=0) -> Y_family_bin
estimateBipartiteSBM(Y_family_bin, model = "bernoulli", dimLabels = c(row = "taxa", col = "soil_code")) -> sbm_family
sbm_family$expectation %>%
  colMeans() %>% sort()

plot(sbm_family, type = 'expected')
```


```{r}

data.frame(sbm_family = sbm_family$memberships$soil_code,
                         soil_code = rownames(Y_family)) %>%
  left_join(soil_df3 %>%
              rownames_to_column("soil_code") %>%
              select(soil_code, coffee_system, Cluster), by ="soil_code")%>%
  left_join(culster_soil_df, by ="soil_code") -> clust_df

B <-
  as.data.frame(
    table(
      clust_df$soil_code,
      clust_df$coffee_system ,
      clust_df$sbm_family,
      clust_df$culster_soil,
      clust_df$Cluster
  )
)


colnames(B) =  c( "soil_code","coffee_system", "sbm_family", "culster_soil", "Cluster","Freq")


w   <- which(B$Freq != 0)
B <- B[w, ]
par(mfrow = c(1,1))
alluvial(B[,c(2,3)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.2,
         cex.axis = 1.5,
         xw = 0.2)

alluvial(B[,c(4,3)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.2,
         cex.axis = 1.5,
         xw = 0.2)
alluvial(B[,c(5,3)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.2,
         cex.axis = 1.5,
         xw = 0.2)
B
NMI(clust_df$coffee_system, clust_df$sbm_family)
NMI(clust_df$culster_soil, clust_df$sbm_family)
NMI(clust_df$Cluster, clust_df$sbm_family)
save.image(file = "outputs/exploration_jsdm.Rdata")
```
