---
title: "Untitled"
output: html_document
date: "2025-04-25"
---

```{r}
library(tidyverse)
library(picante)
library(vegan)
library(ape)
library(readxl)
```

```{r}
data_18S <- "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2"
amf_taxonomic_levels <- c("kingdom", "domain", "phylum", "class", "order", "family",
                          "genus", "species")
amf_taxonomic_levels_eucaryome <- c("kingdom", "phylum", "class", "order", "family",
                          "genus", "species")
metadata <- read_tsv("data/metadata.txt")

amf_OTUs <- data_18S %>%
  #{{import data}}
  read_tsv() %>%
  #{{format column names}}
  rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~sub("_AMF_", "", .)) %>%
  rename_with(~sub("\\_AMF.*", "", .)) %>%
  rename_with(~sub("ROB_T_", "CONT_ROB_", .)) %>%
  rename_with(~sub("ROB_Tex", "CONT_ROB_T_ex", .)) %>%
  mutate(taxonomy = str_replace_all(taxonomy, ":","_"),
       taxonomy = str_replace_all(taxonomy, "[|]","; ")) %>%
  #{{select samples from ROBUST project}}
  select(OTU, amplicon, taxonomy, starts_with("CONT", ignore.case = FALSE),
         starts_with("ROB", ignore.case = FALSE)) %>%
    #{{"Arbuscular mycorrhizal fungi & Mucoromycotina fine root endophytes" are selected}}
  #filter(str_detect(taxonomy,"Glomeromycotina")|str_detect(taxonomy,"Endogonales")) %>%
  filter(str_detect(taxonomy,"p__Glomeromycota")|str_detect(taxonomy,"o__Densosporales")) %>%
  mutate(taxonomy = str_replace_all(taxonomy, "(p__|c__|o__|f__|g__|s__)unclassified", "*"))%>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F) %>%
  #{{select only OTUs and controls}}
  droplevels()

taxonomy <- amf_OTUs %>%
  select(amplicon, OTU, taxonomy) %>%
  separate_wider_delim(taxonomy, names = amf_taxonomic_levels_eucaryome, delim = "; ", cols_remove = F)
 


d <- amf_OTUs %>% 
  replace(amf_OTUs == 0, NA) %>%
  select(amplicon, starts_with("CONT")) %>%
  pivot_longer(-amplicon, names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads))

amf_OTUs_decont <- amf_OTUs %>%
  replace(amf_OTUs == 0, NA) %>%
  pivot_longer(starts_with("ROB"), names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads)) %>%
  #{{merge with control samples}}
  left_join(count(d, amplicon, wt = reads), by = "amplicon") %>%
  #{{substract abundance of control samples}}
  mutate(reads = case_when(
    is.na(n)  ~ reads,
    n > reads ~ 0,
    TRUE      ~ reads - n)) %>%
  select(-n,-starts_with("CONT")) %>%
  pivot_wider(names_from = samples, values_from =  reads, values_fill = 0) %>%
  as_tibble() %>%
  column_to_rownames(var = "amplicon")  
```


# Rarefaction
```{r}
plot.OTUcumSums <-function(x){plot(1:length(colSums(x)), log10(sort(colSums(x))+1))}
amf_OTUs_decont %>%
  select(starts_with("ROB")) %>%
  plot.OTUcumSums
10^3.1
amf_OTUs_decont %>%
  select(starts_with("ROB"))%>%
  colSums()%>%
  sort()


amf_OTUs_decont2 <- amf_OTUs_decont %>%
  #{{sample to exclude}}
  select(starts_with("ROB"))%>%
  select(where(~ sum(.x) >= 10^3))
smallest_amf <- min(colSums(amf_OTUs_decont2))

rarefaction.func <- function(df,rar_sample, n_rare = 100, rarcur = T){
  if(rarcur){ # {{ if rarefaction curve needed}}
    df %>%
      select(starts_with("ROB")) %>%
      t() %>%
      vegan::rarecurve(step = 20, sample = rar_sample, col = "blue", cex = 0.6)
  }
  df %>%
    select(starts_with("ROB")) %>%
    dim -> D
  # {{rarefaction}}
  df_rarefy = matrix(0, nrow =  D[1], ncol = D[2])
  for(n in 1:n_rare){
    df %>%
      select(starts_with("ROB")) %>%
      t() %>%
      vegan::rrarefy(rar_sample) %>%
      t() + df_rarefy -> df_rarefy
    
  }
  round(df_rarefy/n_rare) -> df_rarefy
  df_rarefy %>%
    bind_cols(
      df %>%
        select(!starts_with("ROB"))
    ) -> df_rarefy
  rownames(df_rarefy) = rownames(df)
  return(df_rarefy)
}
amf_OTUs_decont2 %>% 
  mutate(across(starts_with("ROB"), ~as.integer(.x))) -> amf_OTUs_decont2
set.seed(2121349)
rarefaction.func(df = amf_OTUs_decont2, rar_sample = smallest_amf, rarcur = T) -> amf_OTUs_rare
amf_OTUs_rare %>%
  as.data.frame()%>%
  rownames_to_column("amplicon") %>%
  left_join(amf_OTUs_decont %>%
              rownames_to_column("amplicon") %>%
              select(-starts_with("ROB")), by ="amplicon") -> amf_OTUs_rare
```

```{r}

tree_amf <- ape::read.tree(file = "data/18S_AMf_Eukaryome/Robust_MiSeq_18S_AMf_20241120_125_samples_Eukaryome_Glomeromycota.tree")
```

```{r}

soil_df <- read_xlsx("data/soil_data_dry_season_2023.xlsx") %>%
  rename(coffee_system = "Robusta coffee system", soil_code = "SAMPLE NO", sand = "Sand%", clay = "Clay%", silt = "Silt%", pH ="soil pH", 
         OC = "OC%", MO = "OM%", N = "N%", P = "P/mg/kg", K = "K/Cmol/kg", Na = "Na/Cmol/kg") %>%
  mutate(soil_code = str_c("ROB_", soil_code,"_s"))
soil_df %>% 
  filter(soil_code%in% colnames(amf_OTUs_rare))%>%
  mutate(coffee_system = as.factor(coffee_system)) %>%
  column_to_rownames("soil_code")-> soil_df2

```


```{r}
amf_OTUs_rare%>%
  select(-c(ROB_M_s, ROB_L_s)) %>%
  column_to_rownames("amplicon") %>%
  select(starts_with("ROB")) -> Data_abund
str(Data_abund)
Data_abund = as.matrix(Data_abund[tree_amf$tip.label,])

t(t(Data_abund)/rowSums(t(Data_abund))) -> Data_abund_rela
dist(Data_abund_rela) -> afm_abund_rela_dist

rowSums(Data_abund)
apply(Data_abund, 1, which.max) -> OTU_optima_index
as.data.frame(OTU_optima_index) %>% 
  mutate(OTU_optima_sample = colnames(Data_abund)[OTU_optima_index],
         OTU_tot_abund = apply(Data_abund, 1, sum))%>%
  filter(OTU_tot_abund !=0) %>%
  rownames_to_column("OTU") %>%
  left_join(soil_df2%>%
              rownames_to_column("soil_code"), by = join_by("OTU_optima_sample" == "soil_code")) -> optima_df

optima_df %>%
  column_to_rownames("OTU") %>%
  select(pH, N, P,K,Na,MO,sand,clay) %>%
  dist -> dist_optima

drop.tip(tree_amf, setdiff(tree_amf$tip.label, optima_df$OTU)) -> tree_amf_subset
cophenetic(tree_amf_subset) -> amf_phylo_dist

vegan::mantel.correlog(dist_optima,amf_phylo_dist,n.class=50, nperm =999) -> mantel_phylo_optima
plot(mantel_phylo_optima)

Data_abund[optima_df$OTU,] -> Data_abund
dist(Data_abund) -> afm_abund_dist
vegan::mantel.correlog(afm_abund_dist,amf_phylo_dist,n.class=50, nperm =999) -> mantel_phylo
plot(mantel_phylo)

vegan::mantel.correlog(afm_abund_rela_dist,amf_phylo_dist,n.class=50, nperm =999) -> mantel_phylo_rela
plot(mantel_phylo_rela)
```
# NTI

```{r}
t(Data_abund_rela) -> Data_abund_rela_t
MNTD <- picante::ses.mntd(Data_abund_rela_t, cophenetic(tree_amf), abundance.weighted = T, null.model="taxa.labels")
MNTD %>%
  rownames_to_column("soil_code") %>%
  left_join(soil_df2 %>%
              rownames_to_column("soil_code"), by = "soil_code") -> MNTD_df
MNTD_df %>%
  ggplot()+
  geom_histogram(aes(mntd.obs.z))

MNTD_df %>%
  pivot_longer(c(pH,sand,MO,P,K,Na))%>%
  ggplot()+
    facet_wrap(~name, scale ="free")+
    geom_point(aes(value, mntd.obs.z))+
    geom_smooth(aes(value, mntd.obs.z), method = "glm")

MNTD_df %>%
  ggplot()+
    geom_boxplot(aes(coffee_system, mntd.obs.z))
MNTD_df %>%
  ggplot()+
    geom_boxplot(aes(Cluster, mntd.obs.z))

```


# Beta NTI
```{r}
t(Data_abund_rela) -> Data_abund_rela_t
betaMNTD <- picante::comdistnt(Data_abund_rela_t, cophenetic(tree_amf), abundance.weighted = T)
betaMNTD_df <- as.data.frame(as.table(as.matrix(betaMNTD))) %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  dplyr::rename(betaMNTD = Freq)
```



```{r}
library(purrr)
library(furrr)
betaMNTD.null <- function(abund_data, phylo_data, nrand){
  ncol(abund_data) -> ntaxa
  sample(1:ntaxa) -> rand_id
  colnames(abund_data)[rand_id] ->  colnames(abund_data)
  picante::comdistnt(abund_data, phylo_data, abundance.weighted = T) -> betaMNTD_null
  
  as.data.frame(as.table(as.matrix(betaMNTD_null))) %>%
    dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
    dplyr::rename( betaMNTD_null = Freq) -> betaMNTD_null_edgelist
  return(betaMNTD_null_edgelist)
}
plan(multisession, workers = 10)
1:1000%>%
  furrr::future_map(\(x) betaMNTD.null(Data_abund_rela_t, cophenetic(tree_amf), nrand = x),
             .progress = T) -> temp
temp%>%
  bind_rows() %>%
  group_by(Var1,Var2) %>%
  summarise(betaMNTD_null_mean = mean(betaMNTD_null),
            betaMNTD_null_sd = sd(betaMNTD_null)) %>%
  
  left_join(betaMNTD_df, by = join_by("Var1", "Var2"))%>%
  mutate(betaNTI = (betaMNTD -abs(betaMNTD_null_mean)/betaMNTD_null_sd)) -> betaMNTD_df_full

betaMNTD_df_full %>%
  ggplot()+
  geom_histogram(aes(y= betaNTI))

```

```{r}
soil_df2 %>%
  select(pH) %>%
  dist() %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltapH = "Freq") -> deltapH_df
betaMNTD_df_full %>%
  left_join(deltapH_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltapH,betaNTI))+
    geom_smooth(aes(deltapH,betaNTI), method = "glm")

soil_df2 %>%
  select(MO) %>%
  dist() %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltaMO = "Freq") -> deltaMO_df
betaMNTD_df_full %>%
  left_join(deltaMO_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltaMO,betaNTI))+
    geom_smooth(aes(deltaMO,betaNTI), method = "glm")

soil_df2 %>%
  select(Na) %>%
  dist() %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltaNa = "Freq") -> deltaNa_df
betaMNTD_df_full %>%
  left_join(deltaNa_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltaNa,betaNTI))+
    geom_smooth(aes(deltaNa,betaNTI), method = "glm")


```

# Algomerate phylogenetic tree at the family level
```{r}
cluster_family <- amf_OTUs_rare %>%
  unite(taxonomy, kingdom, phylum, class, order, family, sep = "; ",remove = F) %>%
  select(taxonomy, species, amplicon, OTU)
```

```{r}
# 1. Compute full pairwise patristic distance matrix between OTUs
dist_mat <- cophenetic.phylo(tree_amf)  # matrix of tip-to-tip distances

# 2. Prepare cluster mapping
otu_family <- cluster_family %>%
  group_by(taxonomy) %>%
  # if species isn't known agglomerate a the lower taxonomic level known
  mutate(
    family = paste(str_extract(
      taxonomy,"(s_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  ) %>%
  ungroup() %>%
  select(-taxonomy)


# 3. For each pair of species, compute median inter-cluster distance
family_list <- unique(otu_family$family)

# Create a matrix to store median distances
family_dist <- matrix(NA, nrow = length(family_list), ncol = length(family_list),
                       dimnames = list(family_list, family_list))

for (i in seq_along(family_list)) {
  for (j in seq_along(family_list)) {
    if (i <= j) {
      sp1_otus <- otu_family$amplicon[otu_family$family == family_list[i]]
      sp2_otus <- otu_family$amplicon[otu_family$family == family_list[j]]
      
      dists <- dist_mat[sp1_otus, sp2_otus, drop = FALSE]
      family_dist[i, j] <- family_dist[j, i] <- median(dists, na.rm = TRUE)
    }
  }
}

# 4. Create a new tree using median distances â€” convert matrix to phylo
family_tree <- nj(as.dist(family_dist))

# Optional: root or ladderize for plotting
family_tree <- ladderize(family_tree)

# 5. Plot the collapsed tree
plot(family_tree, cex = 0.7, main = "family-level Tree (collapsed from OTUs)")
```

```{r}
group_by(taxonomy) %>%
  # if species isn't known agglomerate a the lower taxonomic level known
  mutate(
    family = paste(str_extract(
      taxonomy,"(s_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_")
  )
Data_abund %>%
  as.data.frame() %>%
  rownames_to_column("amplicon") %>%
  left_join(cluster_family, by ="amplicon") %>%
  select(starts_with("ROB"),taxonomy) %>%
  group_by(taxonomy) %>%
  summarise(across(where(is.numeric),~sum(.x)),
           family = unique(paste(str_extract(
      taxonomy,"(s_[^;*]+$)|([a-z]_[^;*]+)(?=(;\\s*\\*|\\s*\\*|$))"),
      n(),sep="_"))) %>%
  column_to_rownames("family") %>%
  dplyr::select(-taxonomy)-> abund_rela_family

dist(t(abund_rela_family)) -> dist_abund_rela_famiy
dim(cophenetic(family_tree))
dim(as.matrix(dist_abund_rela_famiy))
MNTD <- picante::ses.mntd(t(abund_rela_family), as.matrix(cophenetic(family_tree)), abundance.weighted = T, null.model="taxa.labels")
MNTD %>%
  rownames_to_column("soil_code") %>%
  left_join(soil_df2 %>%
              rownames_to_column("soil_code"), by = "soil_code") -> MNTD_df
MNTD_df %>%
  ggplot()+
  geom_histogram(aes(mntd.obs.z))

MNTD_df %>%
  pivot_longer(c(pH,sand,MO,P,K,Na))%>%
  ggplot()+
    facet_wrap(~name, scale ="free")+
    geom_point(aes(value, mntd.obs.z))+
    geom_smooth(aes(value, mntd.obs.z), method = "glm")

MNTD_df %>%
  ggplot()+
    geom_boxplot(aes(coffee_system, mntd.obs.z))
MNTD_df %>%
  ggplot()+
    geom_boxplot(aes(Cluster, mntd.obs.z))

```


# Beta NTI
```{r}

dim(abund_rela_family)
str(dist_abund_rela_famiy)
str(cophenetic(family_tree))
betaMNTD <- picante::comdistnt(t(abund_rela_family), cophenetic(family_tree), abundance.weighted = T)
betaMNTD_df <- as.data.frame(as.table(as.matrix(betaMNTD))) %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  dplyr::rename(betaMNTD = Freq)
```



```{r}
library(purrr)
library(furrr)
betaMNTD.null <- function(abund_data, phylo_data, nrand){
  ncol(abund_data) -> ntaxa
  sample(1:ntaxa) -> rand_id
  colnames(abund_data)[rand_id] ->  colnames(abund_data)
  picante::comdistnt(abund_data, phylo_data, abundance.weighted = T) -> betaMNTD_null
  
  as.data.frame(as.table(as.matrix(betaMNTD_null))) %>%
    dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
    dplyr::rename( betaMNTD_null = Freq) -> betaMNTD_null_edgelist
  return(betaMNTD_null_edgelist)
}
plan(multisession, workers = 10)
1:1000%>%
  furrr::future_map(\(x) betaMNTD.null(t(abund_rela_family), cophenetic(family_tree), nrand = x),
             .progress = T) -> temp
temp%>%
  bind_rows() %>%
  group_by(Var1,Var2) %>%
  summarise(betaMNTD_null_mean = mean(betaMNTD_null),
            betaMNTD_null_sd = sd(betaMNTD_null)) %>%
  
  left_join(betaMNTD_df, by = join_by("Var1", "Var2"))%>%
  mutate(betaNTI = (betaMNTD -abs(betaMNTD_null_mean)/betaMNTD_null_sd)) -> betaMNTD_df_full

betaMNTD_df_full %>%
  ggplot()+
  geom_histogram(aes(y= betaNTI))

```

```{r}
soil_df2 %>%
  select(pH) %>%
  dist() %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltapH = "Freq") -> deltapH_df
betaMNTD_df_full %>%
  left_join(deltapH_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltapH,betaNTI))+
    geom_smooth(aes(deltapH,betaNTI), method = "glm")

soil_df2 %>%
  select(MO) %>%
  dist() %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltaMO = "Freq") -> deltaMO_df
betaMNTD_df_full %>%
  left_join(deltaMO_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltaMO,betaNTI))+
    geom_smooth(aes(deltaMO,betaNTI), method = "glm")

soil_df2 %>%
  select(Na) %>%
  dist() %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltaNa = "Freq") -> deltaNa_df
betaMNTD_df_full %>%
  left_join(deltaNa_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltaNa,betaNTI))+
    geom_smooth(aes(deltaNa,betaNTI), method = "glm")

soil_df2 %>%
  select(Cluster) %>%
  as.matrix %>%
  as.table %>%
  as.data.frame() %>%
  dplyr::filter(as.character(Var1) < as.character(Var2)) %>%
  rename(deltaNa = "Freq") -> deltaNa_df
betaMNTD_df_full %>%
  left_join(deltaNa_df,by = join_by("Var1", "Var2"))%>%
    ggplot()+
    geom_point(aes(deltaNa,betaNTI))+
    geom_smooth(aes(deltaNa,betaNTI), method = "glm")

```

